### PR TITLE
Package Help dialog updated

### PR DESCRIPTION
**Pull request by CSMathematics**
Updated CTAN package database with more information for each package.
TexDialog

### CODE REVIEW

**Comment by sunderme** [comment]
looks good

* packageList.xml can be removed from qrc and git if we take this solution.
* where is the json database coming from ?
* the dialog is a bit small on my 4k monitor, maybe that can be optimized (especially as it does not remember resized dialogs)

File: src/texdocdialog.h [hunk 1]
```diff
@@ -38,6 +39,7 @@ private slots:
 	QString lastDocRequest;
     Help *help;
     QStringList m_packages;
+    QButtonGroup ButtonGroup;
```
Comment (sunderme) [inline src/texdocdialog.h:42]: to follow (implicit) naming strategy, the variable should start with a lower case letter (buttonGroup or even m_buttonGroup)

File: src/texdocdialog.ui [hunk 2]
```diff
@@ -6,32 +6,72 @@
...
       </property>
-     </column>
+      <widget class="QTextBrowser" name="PackageInfoBrowser"/>
+      <widget class="QWidget" name="">
+       <layout class="QVBoxLayout" name="verticalLayout">
+        <property name="spacing">
+         <number>0</number>
+        </property>
+        <item>
+         <layout class="QHBoxLayout" name="LanguagesLayout">
```
Comment (sunderme) [inline src/texdocdialog.ui:43]: the ususal naming strategy is something like lyLanguages or similar.
lower case first letter helps readability.

File: src/texdocdialog.ui [hunk 3]
```diff
@@ -6,32 +6,72 @@
...
+              <width>40</width>
+              <height>20</height>
+             </size>
+            </property>
+           </spacer>
+          </item>
+         </layout>
+        </item>
+        <item>
+         <widget class="QTextBrowser" name="PackageDescriptions"/>
```
Comment (sunderme) [inline src/texdocdialog.ui:67]: see comment above

File: src/texdocdialog.h (hunk 1 above)
Comment (CSMathematics) [reply src/texdocdialog.h:42]: Good point. I will review the code again to make these changes.

**Comment by CSMathematics** [comment]
The json database is drawn from [CTAN](https://www.ctan.org/help/json/2.0/packages) but it has been edittted.
I merged 4 different json files into 1 database file by building a custom function that gets the data and creating the file. It is not included in this project. Ithink I will be able to upload it in my repository in a few days.

**Comment by CSMathematics** [comment]
The only thing I couldn't solve is to find a way to open the package documentation when the link is clicked inside the QTextBrowser.

**Comment by CSMathematics** [comment]
All changes are made.

**Comment by sunderme** [comment]
does not compile on OSX (LLVM)

File: src/texdocdialog.cpp [hunk 4]
```diff
@@ -140,7 +143,28 @@ void TexdocDialog::itemChanged(QTableWidgetItem* item)
...
+    buttonGroup.setExclusive(true);
+
+    while(buttonGroup.buttons().count()>0){
+        QAbstractButton * bt = buttonGroup.buttons().first();
+        ui->LanguagesLayout->removeWidget(bt);
+        buttonGroup.removeButton(bt);
+        delete bt;
+    }
+    for(CTANDescription description : package.descriptions){
+        QPushButton * LangButton = new QPushButton(description.language,this);
```
Comment (sunderme) [inline src/texdocdialog.cpp:159]: *L*angbutton

File: src/texdocdialog.cpp [hunk 5]
```diff
@@ -140,7 +143,28 @@ void TexdocDialog::itemChanged(QTableWidgetItem* item)
...
-	delayedCheckDocAvailable(text);
+    LatexPackageInfo package = item->data(Qt::UserRole).value<LatexPackageInfo>();
+    QString Info = LatexRepository::packageInfo(package);
+    ui->packageInfoBrowser->setMarkdown(Info);
+    delayedCheckDocAvailable(text);
+    buttonGroup.setExclusive(true);
+
+    while(buttonGroup.buttons().count()>0){
+        QAbstractButton * bt = buttonGroup.buttons().first();
+        ui->LanguagesLayout->removeWidget(bt);
```
Comment (sunderme) [inline src/texdocdialog.cpp:154]: LanguageLayout should start with lower case

Comment (CSMathematics) [reply src/texdocdialog.cpp:154]: Missed that. I will fix it right away

**Comment by CSMathematics** [comment]
> does not compile on OSX (LLVM)

**Comment by CSMathematics** [comment]
> > does not compile on OSX (LLVM)

I changed the variable names. But I can't see the reason that compilation fails on OSX. I looked at the compile log and if I'm not mistaken the errors are on other files than the ones I modified.

**Comment by sunderme** [comment]
thanks

**Comment by tmelorc** [comment]
There are blank spaces before the colon. 
For example, `Name : a0poster`

**Comment by octaeder** [comment]
As discussed with latex help and txs manual, it is preferred to hold sources of information locally. So the question is what happens here when you have no internet connection? Can txs handle this? Is the list of packages including captions still available? Has anyone tested this case?

**Comment by tmelorc** [comment]
In addition, I tried to compile it now:
```
-- The current git hash is a8c0199
-- The current git describe is 4.4.0-1053-ga8c01995f
```

But I got the error:
Seleção_101

**Comment by octaeder** [comment]
I was able to build with msys2 ucrt64

**Comment by octaeder** [comment]
TeXstudio 4.7.3 (git 4.7.2-38-ga8c01995f)
Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

**Comment by tmelorc** [comment]
> TeXstudio 4.7.3 (git 4.7.2-38-ga8c01995f) Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

It seems that `setMarkdown` is available only for Qt6.6.

**Comment by octaeder** [comment]
Is this the same?
https://doc.qt.io/qt-6.2/qtextdocument.html#setMarkdown
https://doc.qt.io/qt-5/qtextdocument.html#setMarkdown

**Comment by tmelorc** [comment]
> Is this the same? https://doc.qt.io/qt-6.2/qtextdocument.html#setMarkdown https://doc.qt.io/qt-5/qtextdocument.html#setMarkdown

Looks like. But the doc says _since 5.14_. Let me check my Qt version. Maybe I have older one.

**Comment by dbitouze** [comment]
Nice feature!

But I don't understand the point of by default disabling the search text field: as long as the “all the packages” option is disabled (the corresponding checkbox is unchecked), nothing can be done. What I would expect would be that, by default:

- the “all the packages” option is disabled
- the search text field is enabled, allowing the user to search for a particular package (without cluttering the interface with the list of all packages).

**Comment by octaeder** [comment]
@dbitouze this is impossible for you?

image

and I would say that the option is always disabled when the dialog is started

**Comment by tmelorc** [comment]
I'm using the version below and the text field is locked, as screenshot below.
```
TeXstudio 4.7.3 (git 4.7.1-81-gf0f4089a9)
Utilizando Qt Versão 5.15.3, compilado com Qt 5.15.3 R

```
Documentação dos Pacotes (Texdoc)_002

**Comment by octaeder** [comment]
Including 5 extra developments and missing the latest 2 (reagarding cwls) I use this (msys2 build) on Win10:

TeXstudio 4.7.3 (git 4.7.2-51-gbd8a0bd91)
Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

I wonder why you see no packages at all. In any case the packages list should not be empty at start. How do you get the packages into place?

**Comment by tmelorc** [comment]
> I wonder why you see no packages at all. In any case the packages list should not be empty at start. How do you get the packages into place?

Sorry, I didn't follow the question.

**Comment by tmelorc** [comment]
The packages list only appears after the check button is on. Then the text field is unlocked.

Documentação dos Pacotes (Texdoc)_003

**Comment by octaeder** [comment]
there's something wrong.

**Comment by tmelorc** [comment]
> there's something wrong.

Maybe this line: `ui->lineEditSearch->setEnabled(n>0);`

**Comment by octaeder** [comment]
since search field is not enabled I would conclude that n equals 0. But then m_packages is empty & state<=0. Since I can't reproduce, can can you debug the the values of m_packages.count() and state immediately after line 48 (int n=0;)?

**Comment by octaeder** [comment]
btw: do you use packages in your latex document?

**Comment by tmelorc** [comment]
> btw: do you use packages in your latex document?

You got it! I was opening the texdoc window with no opened document.
Now it works.

Anyway, it would be good to have the feature available also for no documents.

**Comment by octaeder** [comment]
in this case the option should be checked

**Pull request merged**